soroban-spec/src/gen/ts.rs

```diff
@@ -0,0 +1,339 @@
+use std::{fs, io};
+
+use super::json::types::{self, StructField, UnionCase};
+
+use heck::ToLowerCamelCase;
+use itertools::Itertools;
+use sha2::{Digest, Sha256};
+use stellar_xdr::ScSpecEntry;
+
+use types::Entry;
+
+use crate::{
+    gen::{json::types::Type, ts::wrapper::type_to_js_xdr},
+    read::{from_wasm, FromWasmError},
+};
+
+pub mod boilerplate;
+pub mod wrapper;
+
+#[derive(thiserror::Error, Debug)]
+pub enum GenerateFromFileError {
+    #[error("reading file: {0}")]
+    Io(io::Error),
+    #[error("sha256 does not match, expected: {expected}")]
+    VerifySha256 { expected: String },
+    #[error("parsing contract spec: {0}")]
+    Parse(stellar_xdr::Error),
+    #[error("getting contract spec: {0}")]
+    GetSpec(FromWasmError),
+}
+
+pub fn generate_from_file(
+    file: &str,
+    verify_sha256: Option<&str>,
+) -> Result<String, GenerateFromFileError> {
+    // Read file.
+    let wasm = fs::read(file).map_err(GenerateFromFileError::Io)?;
+
+    // Produce hash for file.
+    let sha256 = Sha256::digest(&wasm);
+    let sha256 = format!("{sha256:x}");
+
+    if let Some(verify_sha256) = verify_sha256 {
+        if verify_sha256 != sha256 {
+            return Err(GenerateFromFileError::VerifySha256 { expected: sha256 });
+        }
+    }
+
+    // Generate code.
+    let json = generate_from_wasm(&wasm).map_err(GenerateFromFileError::GetSpec)?;
+    Ok(json)
+}
+
+pub fn generate_from_wasm(wasm: &[u8]) -> Result<String, FromWasmError> {
+    let spec = from_wasm(wasm)?;
+    let json = generate(&spec);
+    Ok(json)
+}
+
+pub fn generate(spec: &[ScSpecEntry]) -> String {
+    let collected: Vec<_> = spec.iter().map(Entry::from).collect();
+    collected.iter().map(entry_to_ts).join("\n")
```

Comment:
I'm surprised the way we generate TS is by converting from `ScSpecEntry` to `json::Entry` to TS. The JSON format outputs a contract interface as JSON, not a contract client, or even usable types by a contract. It serves a different purpose. I wonder if this will come back to bite us, as we may want to change the format of the JSON interface outputted without changing the output of the TS client. `ScSpecEntry` canonically describes the contract interface and contains the most information. Given it's available I expected we'd convert directly from it.

Reply:
JSON seemed close enough for converting to TS types. What other info is lost?

Personally I think we should write a XDR -> JSON schema parser. This way the generated schema is self describing and provide an additional validation method and make use of tools that generate a form from json schema.